src/mlir/design.md

```diff
@@ -3,69 +3,225 @@
-Adding two numbers (`a + b`) is the same as calling the method `+` on `a` over `b` (`a.+(b)`), but these methods don't exist (how would we implement addition if not with a `+`?).
+## Type Constraints
 
-So the generator has to find those that weren't declared and replace with arithmetic if the types are compatible (int, float).
+Eventually, all types will be clear on the AST, but this isn't the case right now.
+For now, we have to add truncations and extensions where it isn't clear.
 
-Users are still allowed to use `+` as a method name, in which case the generator will recognise and call that method instead of transforming into arithmetic, even if the types are compatible.
 For example:
+ * Constants of type `int` or `float` are created as 64-bit numbers and then truncated if assigned to 32-bit values.
+ * Arithmetic between types of different width have the smallest type promoted to the largest.
 
-```
-// Verona code
-class U32 {}
-class Foo {
-  static +(a: U32, b: U32) { ... }
-}
-main() : U32 {
-  let a = 21 + 21; // This is simple arithmetic
-  let b = Foo::+(21, 21); // This calls the method
-}
-```
+Also, due to a constraint on the standard dialect in MLIR, we currently only support sign-less integers (neither signed nor unsigned).
+Arithmetic will be lowered in the LLVM dialect, so will contain full support for signed and unsigned as well as all intrisics.
 
-## Type Constraints
+We also currently delay stack allocation until the type is defined (node `oftype`), because we don't yet have an idea of the type of local definitions in their own nodes.
+This makes for cumbersome pattern of:
+ 1. Adding an empty value on the symbol table (to reserve the name in the table)
+ 2. Allocating enough space when we see a type declaration (now the value is the address of the new memory)
+ 3. Accessing the address via references to the variable name (map lookup)
+
+Once more nodes have types (ex: `let`, `var`) we can skip the second step.
+
+## Type capabilities and regions
+
+The three main type capabilities are: `imm`, `mut` and `iso`. See https://microsoft.github.io/verona/explore.html#regions.
+
+Immutable regions will always be allocated on the same reference counted "region", via runtime calls.
+They can only be read from, so there are no concurrency issues. They can be collected when there are no more references to them.
+
+Mutable regions will be allocated on the heap, but cannot be used directly, unless they're in a "forest", accessible through their roots (`iso`s).
+
+`Cown` root will be scheduled via the `when` keywords and give no guarantee of completion time or order.
+The runtime calls will be just to create the `cown` and to push it into the scheduler.
+
+**QUESTION: Do stack roots also push behaviours to the scheduler or do they execute synchronously?**
+Either way, they should also yield similar runtime calls.
+
+Regions can also have two more properties:
```

Comment:
At least two more properties.  I suspect we will end up with a complex matrix of properties eventually.